src/loader/resolvable-promise.js

```diff
@@ -0,0 +1,84 @@
+/**
+ * A Promise with resolve() and reject() methods so that it can be fulfilled
+ * asynchronously from outside instead of the callback.
+ *
+ * Because Promise cannot be extended in the usual way, we need to do some
+ * magic to define this class.
+ */
+export default class ResolvablePromise {
+	/**
+	 * Don't construct ResolvablePromise objects directly: rely on
+	 * ResolvablePromise.new() instead.
+	 */
+	constructor() {
+		throw new Error(
+			'Don\'t construct ResolvablePromise objects directly: ' +
+				'rely on ResolvablePromise.new() instead'
+		);
+	}
+}
+
+ResolvablePromise.new = () => {
+	const capture = {};
+
+	const promise = new Promise((resolve, reject) => {
+		capture._resolve = resolve;
+		capture._reject = reject;
+	});
+
+	Object.assign(promise, capture, {
+		fulfilled: false,
+		resolved: false,
+		rejected: false,
+		resolution: undefined,
+		rejection: undefined,
+	});
```

Comment:
Again, because this class is somewhat strange (it tries to mimic a regular class but that can be done because extending Promise directly does not work) I tried to give hints as to what gets exported in the final ResolvablePromise and what doesn't.

I mean, all ResolvablePromise methods are static functions that receive the Promise as the first argument, but some of them are exported through object fields:

![imagen](https://user-images.githubusercontent.com/3273630/52043991-4effce00-2541-11e9-9a06-5f4856ea0bf3.png)

And some others don't. And that's why I added the `_` hint to the function name. In the end, all the methods are private to the module (even resolve and reject). They are accessible only because of the "export" I pasted above.